packages/libs/error-stack/src/fmt/mod.rs

```diff
@@ -0,0 +1,1075 @@
+#[cfg(feature = "unstable")]
+mod unstable;
+
+use alloc::{
+    borrow::ToOwned,
+    collections::VecDeque,
+    format,
+    string::{String, ToString},
+    vec,
+    vec::Vec,
+};
+use core::{
+    fmt,
+    fmt::{Debug, Display, Formatter},
+    mem,
+};
+
+#[cfg(feature = "std")]
+pub use hook::builtin;
+#[cfg(not(feature = "std"))]
+#[allow(clippy::redundant_pub_crate)]
+pub(crate) use hook::builtin;
+#[cfg(feature = "std")]
+pub use hook::HookContext;
+use hook::HookContextImpl;
+#[cfg(feature = "std")]
+pub(crate) use hook::Hooks;
+#[cfg(feature = "pretty-print")]
+use owo_colors::{OwoColorize, Stream::Stdout, Style as OwOStyle};
+#[cfg(feature = "unstable")]
+pub use unstable::DebugDiagnostic;
+
+use crate::{AttachmentKind, Context, Frame, FrameKind, Report};
+
+/// Modify the behaviour, with which text returned from hook invocations are rendered.
+///
+/// Text can either be emitted immediately as the [`Emit::Next`] line, or deferred the via
+/// [`Emit::Defer`] until the end of the current stack, a stack is a list of attachments until a
+/// frame, which has more than a single source.
+///
+/// Hooks that emit [`Emit::Defer`] are reversed when rendered, meaning that when `A`, `B`, and `C`
+/// use [`Emit::Defer`], they will be rendered in the order: `C`, `B`, `A`.
+///
+/// # Example
+///
+/// ```rust
+/// # // we only test on nightly, therefore report is unused (so is render)
+/// # #![cfg_attr(not(nightly), allow(dead_code, unused_variables, unused_imports))]
+/// use std::io::{Error, ErrorKind};
+///
+/// use error_stack::{
+///     fmt::Emit,
+///     Report
+/// };
+///
+/// Report::install_debug_hook::<u64>(|val, _| Emit::next(format!("u64: {val}")));
+/// Report::install_debug_hook::<u32>(|val, _| Emit::defer(format!("u32: {val}")));
+///
+/// let report = Report::new(Error::from(ErrorKind::InvalidInput))
+///     .attach(1u64)
+///     .attach(2u32)
+///     .attach(3u64)
+///     .attach(4u32)
+///     .attach(5u32)
+///     .attach(6u64)
+///     .attach(7u64);
+///
+/// # owo_colors::set_override(true);
+/// # fn render(value: String) -> String {
+/// #     let backtrace = regex::Regex::new(r"Backtrace No\. (\d+)\n(?:  .*\n)*  .*").unwrap();
+/// #     let backtrace_info = regex::Regex::new(r"backtrace with (\d+) frames \((\d+)\)").unwrap();
+/// #
+/// #     let value = backtrace.replace_all(&value, "Backtrace No. $1\n  [redacted]");
+/// #     let value = backtrace_info.replace_all(value.as_ref(), "backtrace with [n] frames ($2)");
+/// #
+/// #     ansi_to_html::convert_escaped(value.as_ref()).unwrap()
+/// # }
+/// #
+/// # #[cfg(nightly)]
+/// # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__emit.snap")].assert_eq(&render(format!("{report:?}")));
+/// #
+/// println!("{report:?}");
+/// ```
+///
+/// <pre>
+#[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__emit.snap"))]
+/// </pre>
+#[derive(Debug, Clone)]
+#[must_use]
+pub enum Emit {
```

Comment:
Just in case, do we want to mark it `#[non_exhaustive]`?

Reply:
added in 93a150aa4c1193cae65f69f3e877034c1faf63e9